src/metrics/include/metrics/v2/event.hpp

```diff
@@ -247,70 +292,274 @@ class Collector {
+    bool _assertMetricsBuffer;
+    std::thread _thread;
+    std::set<StreamPtr> streams;
+    std::mutex _streamsMutex;
+};
+
+// Manages all the grpc threads. Divides the workload evenly between them.
+template <typename ClockSource, typename StreamInterface>
+class GrpcClient {
+public:
+    GrpcClient(bool assertMetricsBuffer) {
+        for (int i = 0; i < CLIENT_THREADS; i++) {
+            _threads.emplace_back(assertMetricsBuffer);
+        }
+    }
+
+    // EventStream users manage their own memory and should deregister themselves before
+    // destructing, so we (carefully) use naked pointers.
+    typedef EventStream<ClockSource, StreamInterface>* StreamPtr;
+
+    void registerStream(StreamPtr stream) {
+        _threads[latest_thread++ % _threads.size()].registerStream(stream);
+    }
+
+    void closeStream(StreamPtr stream) {
+        for (int i = 0; i < _threads.size(); i++) {
+            _threads[i].closeStream(stream);
+        }
+    }
+
+private:
+    // deque avoid copy-constructor calls
+    std::deque<GrpcThread<ClockSource, StreamInterface>> _threads;
+    std::atomic<size_t> latest_thread = 0;
+};
+
+template <typename ClockSource>
+struct MetricsArgs {
+    MetricsArgs(const typename ClockSource::time_point& finish,
+                OperationEventT<ClockSource> event,
+                size_t workerCount)
+        : finish{finish}, event{std::move(event)}, workerCount{workerCount} {}
+    typename ClockSource::time_point finish;
+    OperationEventT<ClockSource> event;
+    size_t workerCount;
+};
+
+// Efficient buffer that should rarely be blocked on insertions.
+template <typename ClockSource>
+class MetricsBuffer {
+public:
+    explicit MetricsBuffer(size_t size, const std::string& name)
+        : size{size},
+          name{name},
+          _loading{std::make_unique<std::vector<MetricsArgs<ClockSource>>>()},
+          _draining{std::make_unique<std::vector<MetricsArgs<ClockSource>>>()} {
+        _draining->reserve(size);
+        _loading->reserve(size);
+    }
+
+    // Thread-safe.
+    void addAt(const typename ClockSource::time_point& finish,
+               OperationEventT<ClockSource> event,
+               size_t workerCount) {
+        const std::lock_guard<std::mutex> lock(_loadingMutex);
+        _loading->emplace_back(finish, std::move(event), workerCount);
+    }
+
+    // Not thread-safe.
+    std::optional<MetricsArgs<ClockSource>> pop(bool force, bool assertMetricsBuffer=true) {
+        refresh(force, assertMetricsBuffer);
+        if (_draining->empty()) {
+            return std::nullopt;
+        } else {
+            auto ret = std::move(_draining->back());
+            _draining->pop_back();
+            return ret;
+        }
+    }
+
+    const std::string name;
+    const size_t size;
+
+private:
+    void refresh(bool force, bool assertMetricsBuffer) {
+        if (_draining->empty()) {
+            const std::lock_guard<std::mutex> lock(_loadingMutex);
```

Comment:
super nit-picky (too much coffee this am apparently):
I prefer to have early-return rather than the full impl inside the `if` block 

```
if (!draining) {
  return;
}
...rest of fn...
```